web-registry/src/hooks/projects/useProjectsWithOrders.ts

```diff
@@ -0,0 +1,101 @@
+import { SellOrderInfo } from '@regen-network/api/lib/generated/regen/ecocredit/marketplace/v1/query';
+import {
+  ProjectInfo,
+  QueryProjectsResponse,
+} from '@regen-network/api/lib/generated/regen/ecocredit/v1/query';
+import shuffle from 'lodash/shuffle';
+
+import {
+  ProjectsSellOrders,
+  useProjectsSellOrders,
+} from 'pages/Projects/hooks/useProjectsSellOrders';
+import useEcocreditQuery from 'hooks/useEcocreditQuery';
+import { useQuerySellOrders } from 'hooks/useQuerySellOrders';
+
+const hasMetadata = (project: ProjectInfo): boolean => !!project.metadata;
+
+const hasSellOrders = (
+  project: ProjectInfo,
+  sellOrders: SellOrderInfo[],
+): boolean =>
+  sellOrders.some(sellOrder => sellOrder?.batchDenom?.startsWith(project.id));
+
+const prioritizeWithSellOrders = (
+  projects: ProjectInfo[],
+  sellOrders: SellOrderInfo[],
+  limit?: number,
+): ProjectInfo[] => {
+  const projectsWithOrders = projects.filter(projs =>
+    hasSellOrders(projs, sellOrders),
+  );
+  if (!limit || projectsWithOrders.length >= limit) return projectsWithOrders;
+
+  const numProjects = limit - projectsWithOrders.length;
+  const additionalProjects = projects.filter(
+    p1 => !projectsWithOrders.some(p2 => p1.id === p2.id),
+  );
+  return [...projectsWithOrders, ...additionalProjects.slice(0, numProjects)];
+};
+
+const isOtherProject = (project: ProjectInfo, projectId: string): boolean =>
+  project.id !== projectId;
+
+interface ProjectsWithOrdersProps {
+  limit?: number;
+  metadata?: boolean; // to discard projects without metadata prop
+  random?: boolean; // to shuffle the projects (along with limit allows a random subselection)
+  projectId?: string; // to discard an specific project
+}
+
+/**
+ * Get projects and sell orders and merge them both normalizing
+ */
+export function useProjectsWithOrders({
+  limit,
+  metadata = false,
+  random = false,
+  projectId,
+}: ProjectsWithOrdersProps): ProjectsSellOrders {
+  const { data, loading: loadingProjects } =
+    useEcocreditQuery<QueryProjectsResponse>({
+      query: 'projects',
+      params: {},
+    });
+  const projects = data?.projects;
+
+  const { sellOrdersResponse } = useQuerySellOrders();
+  const sellOrders = sellOrdersResponse?.sellOrders;
+
+  const { projectsWithOrderData, loading: loadingWithOrders } =
+    useProjectsSellOrders({
+      projects: selectProjects(projects, sellOrders),
+      sellOrders,
+      limit,
+    });
+
+  function selectProjects(
+    _projects: ProjectInfo[] | undefined,
+    sellOrders: SellOrderInfo[] | undefined,
+  ): ProjectInfo[] | undefined {
+    if (!_projects || !sellOrders) return;
+
+    if (projectId)
+      _projects = _projects.filter(proj => isOtherProject(proj, projectId));
+
+    if (random)
+      return prioritizeWithSellOrders(
```

Comment:
I don't think calling this function is needed anyways. Once the projects have been shuffled + filtered on metadata, they will be sorted again on sell orders here: https://github.com/regen-network/regen-web/blob/master/web-registry/src/pages/Projects/hooks/useProjectsSellOrders.tsx#L59

Sorting after shuffling should give you the order you want I think. Try to remove it, should give you the same result.

Reply:
hmm, but the thing is we don't want to sort after shuffling. We want 3 random projects, prioritizing those with sell orders... (at the time of implementing this only 2 projects with sell orders in redwood)

Reply:
Yes, that was my point. If you shuffle an array and then sort it on a field, it will still be like picking random projects but with having the one with the field sorted first.
![image](https://user-images.githubusercontent.com/530644/189839856-d6c5233e-d1d6-4a27-8bc4-aa565b81f21a.png)
When you're writing a sorting function you're not necessarily changing all the element's order.
In this case, we can just put elements with sell order first and leave the rest in the same place in the array.

This is what this sort function is doing: https://github.com/regen-network/regen-web/blob/a1e5bb31afd475adebe8f4a4ab7026a58422a6b2/web-registry/src/pages/Projects/hooks/useProjectsSellOrders.utils.tsx#L74

Reply:
The only thing that might be needed on top of shuffling the projects is removing the current one I think.

Reply:
ok, I was not aware that sortProjectsBySellOrdersAvailability ordering was recently introduced in the final step in getProjectDisplayData

Reply:
@flagrede Done! 
I've also fixed the `sortProjectsBySellOrdersAvailability` function, as it was putting projects with sell orders at the end instead of at the beginning. I have also improved the performance by changing `filter` to `some` since it is enough for us to know that a project has a sell order, right?

Reply:
Sounds good, weird that it was reversed but I trust you on this.